qiskit_experiments/curve_analysis/drawer_mixin.py

```diff
@@ -0,0 +1,512 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Curve drawing functionality for analysis."""
+
+import functools
+from typing import Callable, Set, List, Dict, Sequence, Union, Optional
+
+import numpy as np
+from matplotlib.axes import Axes
+from matplotlib.ticker import FuncFormatter, ScalarFormatter
+from qiskit.utils import detach_prefix
+from uncertainties import unumpy as unp, UFloat
+
+from qiskit_experiments.curve_analysis.curve_data import FitData
+from qiskit_experiments.framework import AnalysisResultData, Options
+from qiskit_experiments.framework.matplotlib import get_non_gui_ax
+
+
+class CurveDrawerMixin:
+    """Mixin to provide drawing functionalities for curve fit result."""
+
+    # Internal state for drawing options
+    _draw_options: Options
+    _set_draw_options: Set
+
+    def __init_subclass__(cls, **kwargs):
+        """Initialize drawing options."""
+        super().__init_subclass__(**kwargs)
+
+        # This is a hack to let the mixin have internal state for drawing option.
+        @functools.wraps(cls.__init__, assigned=("__annotations__",))
+        def __new__(cls, *args, **kwargs):
+            instance = super().__new__(cls, *args, **kwargs)
+            setattr(instance, "_default_draw_options", cls._default_draw_options)
+            setattr(instance, "_set_draw_options", set())
+            return instance
+
+        cls.__new__ = __new__
```

Comment:
Can you elaborate on what is going on here? Is this to suppress lint warnings?

Reply:
This is to completely separate the context of drawer from the curve analysis, though we need a state for draw options. Basically mixin class should be stateless. 

```
setattr(instance, "_default_draw_options", cls._default_draw_options)
```
This line was my mistake. This is correct method to add state, i.e. draw option, to curve analysis instance from the mixin. 

```
        def __new__(sub_cls, *args, **kwargs):
            instance = super(cls, sub_cls).__new__(sub_cls, *args, **kwargs)
            instance._draw_options = instance._default_draw_options()
            instance._set_draw_options = set()

            return instance
```

This complex overhead is due to readability. If we initialize `self. _draw_options ` in `CurveAnalysis` where it doesn't define `_default_draw_options ` class method, it may look weird. By doing this, the developer of curve fitting module is completely agnostic to drawer (unless one updates the default draw options).

The reason that I separated the options is due to extendability and usability. Current default curve analysis options are already too much complicated 
https://github.com/Qiskit/qiskit-experiments/blob/8f035af5fdf7e9fcbe66b8f071327085f6b1ba8f/qiskit_experiments/curve_analysis/curve_analysis.py#L351-L371
note that `options.style` has further nested options. This separation will simplify the structure and allow user to easily find target option name. 

Alternative approach would be defining drawer as an instance, rather than mixin.
```python
class CurveAnalysis(BaseAnalysis):

    def __init__(self, drawer: CurveDrawer):
        self.drawer = drawer
```
then we can internally call drawing functions as something like `self.drawer._draw_fit_lines(...)` rather than `self._draw_fit_lines(...)`. The drawback is it moves draw options one layer deeper, i.e. `my_exp.analysis.drawer.set_options()` so it might hurt usability. Another negative effect would be serialization. Since all constructor args should be serializable, we need to implement json and pickle logic for the drawer class.

Reply:
Base on our offline discussion I decided to change drawer from mixin to instance.